File: c++/knuth_shuffle/knuth_shuffle.cpp
#include <stdio.h>
#include "knuth_shuffle.h"

int main()
{
  srand(time(NULL));

  int aryLen = 10;
  int ary[aryLen];
  for (int i = 0; i < aryLen; i++) {
    ary[i] = i;
    printf("%d ", ary[i]);
  }
  printf("\n");

  knuthShuffle(&ary[0], aryLen);

  for (int i = 0; i < aryLen; i++) {
    printf("%d ", ary[i]);
  }
  printf("\n");
}

File: ruby/knuth_shuffle/knuth_shuffle.rb
def knuth_shuffle(ary)
  (1..ary.size).each { |i|
    outer = ary.size - i
    swap = rand(outer)
    temp = ary[swap]
    ary[swap] = ary[outer]
    ary[outer] = temp
  }
end

aryLen = 10;
ary = [*0...aryLen]
puts ary.to_s
knuth_shuffle(ary)
puts ary.to_s
